File: java/basic_app/app/src/main/java/com/appsflyer/onelink/appsflyeronelinkbasicapp/ConversionDataActivity.java
package com.appsflyer.onelink.appsflyeronelinkbasicapp;

import androidx.appcompat.app.AppCompatActivity;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Map;
import java.util.SortedMap;
import java.util.SortedSet;
import java.util.TreeSet;

public class ConversionDataActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_conversion_data);
        Map<String, Object> conversionData = ((AppsflyerBasicApp) getApplication()).conversionData;
        displayConversionData(conversionData);
    }

    public void displayConversionData(Map<String, Object> conversionData){
        String conversionDataString = mapToSortedString(conversionData);
        TextView conversionDataTextView = findViewById(R.id.conversionDataTextView);
        conversionDataTextView.setMovementMethod(new ScrollingMovementMethod());
        conversionDataTextView.setText(conversionDataString);
    }

    public String mapToSortedString(Map<String, Object> map){
        String result = "";
        SortedSet<String> keys = new TreeSet<>(map.keySet());
        Object value;
        for (String key: keys){
            value = map.get(key);
            if (value == null) {
                value = "null";
            }
            result += String.format("%s : %s\n", key, value);
        }
        return result;
    }
}

File: kotlin/basic_app/app/src/main/kotlin/com/appsflyer/onelink/appsflyeronelinkbasicapp/ConversionDataActivity.kt
package com.appsflyer.onelink.appsflyeronelinkbasicapp

import android.os.Bundle
import android.text.method.ScrollingMovementMethod
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class ConversionDataActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_conversion_data)
        val conversionData = (application as AppsflyerBasicApp).conversionData
        displayConversionData(conversionData)
    }

    //A function that receives a map containing the conversion data and displays it
    private fun displayConversionData(conversionData: Map<String, Any>?) {
        val conversionDataString = mapToSortedString(conversionData)
        val conversionDataTextView = findViewById<TextView>(R.id.conversionDataTextView)
        conversionDataTextView.movementMethod = ScrollingMovementMethod()
        conversionDataTextView.setText(conversionDataString)

    }
// A function that accepts a Map variable and converts it to a String variable
    private fun mapToSortedString(map: Map<String, Any>?): String {
        var result = ""

        map?.let {
            for (keyValuePair in map) {
                result += "${keyValuePair.key} : ${keyValuePair.value}\n"
            }
        } ?: run {
            result = "Conversion data not available at the moment"
        }
        return result
    }

}